// @motion-canvas/2d/lib/components/View2D.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Rect } from './Rect';
import { initial, signal } from '../decorators';
import { PlaybackState } from '@motion-canvas/core';
import { lazy } from '@motion-canvas/core/lib/decorators';
export class View2D extends Rect {
    constructor(props) {
        super({
            composite: true,
            fontFamily: 'Roboto',
            fontSize: 48,
            lineHeight: '120%',
            textWrap: false,
            fontStyle: 'normal',
            ...props,
        });
        this.view2D = this;
        View2D.shadowRoot.append(this.element);
        this.applyFlex();
    }
    transformContext() {
        // do nothing
    }
    dispose() {
        this.removeChildren();
        super.dispose();
    }
    render(context) {
        this.computedSize();
        this.computedPosition();
        super.render(context);
    }
    requestLayoutUpdate() {
        this.updateLayout();
    }
    view() {
        return this;
    }
}
__decorate([
    initial(PlaybackState.Paused),
    signal()
], View2D.prototype, "playbackState", void 0);
__decorate([
    lazy(() => {
        const frameID = 'motion-canvas-2d-frame';
        let frame = document.querySelector(`#${frameID}`);
        if (!frame) {
            frame = document.createElement('div');
            frame.id = frameID;
            frame.style.position = 'absolute';
            frame.style.pointerEvents = 'none';
            frame.style.top = '0';
            frame.style.left = '0';
            frame.style.opacity = '0';
            frame.style.overflow = 'hidden';
            document.body.prepend(frame);
        }
        return frame.shadowRoot ?? frame.attachShadow({ mode: 'open' });
    })
], View2D, "shadowRoot", void 0);
