// @motion-canvas/core/lib/signals/CompoundSignalContext.js
import { map } from '../tweening';
import { SignalContext } from './SignalContext';
import { isReactive } from './isReactive';
export class CompoundSignalContext extends SignalContext {
    constructor(entries, parser, initial, interpolation, owner = undefined) {
        super(undefined, interpolation, owner);
        this.entries = entries;
        this.signals = [];
        this.parser = parser;
        for (const key of entries) {
            const signal = new SignalContext(isReactive(initial)
                ? () => parser(initial())[key]
                : parser(initial)[key], map, owner ?? this.invokable).toSignal();
            this.signals.push([key, signal]);
            Object.defineProperty(this.invokable, key, { value: signal });
        }
    }
    toSignal() {
        return this.invokable;
    }
    parse(value) {
        return this.parser(value);
    }
    get() {
        return this.parse((Object.fromEntries(this.signals.map(([key, property]) => [key, property()]))));
    }
    set(value) {
        if (isReactive(value)) {
            for (const [key, property] of this.signals) {
                property(() => this.parser(value())[key]);
            }
        }
        else {
            const parsed = this.parse(value);
            for (const [key, property] of this.signals) {
                property(parsed[key]);
            }
        }
        return this.owner;
    }
    reset() {
        for (const [, signal] of this.signals) {
            signal.reset();
        }
        return this.owner;
    }
    save() {
        for (const [, signal] of this.signals) {
            signal.save();
        }
        return this.owner;
    }
}
