// @motion-canvas/core/lib/signals/DependencyContext.js
import { FlagDispatcher } from '../events';
import { DetailedError } from '../utils';
export class DependencyContext {
    static collectPromise(promise, initialValue = null) {
        const handle = {
            promise,
            value: initialValue,
            stack: new Error().stack,
        };
        const context = this.collectionStack.at(-1);
        if (context) {
            handle.owner = context.owner;
        }
        promise.then(value => {
            handle.value = value;
            context?.markDirty();
        });
        this.promises.push(handle);
        return handle;
    }
    static consumePromises() {
        const result = this.promises;
        this.promises = [];
        return result;
    }
    constructor(owner) {
        this.owner = owner;
        this.dependencies = new Set();
        this.event = new FlagDispatcher();
        this.markDirty = () => this.event.raise();
        this.invokable = this.invoke.bind(this);
        Object.defineProperty(this.invokable, 'context', {
            value: this,
        });
        Object.defineProperty(this.invokable, 'toPromise', {
            value: this.toPromise.bind(this),
        });
    }
    invoke() {
        // do nothing
    }
    startCollecting() {
        if (DependencyContext.collectionSet.has(this)) {
            throw new DetailedError('A circular dependency occurred between signals.', `This can happen when signals reference each other in a loop.
        Try using the attached stack trace to locate said loop.`);
        }
        DependencyContext.collectionSet.add(this);
        DependencyContext.collectionStack.push(this);
    }
    finishCollecting() {
        DependencyContext.collectionSet.delete(this);
        if (DependencyContext.collectionStack.pop() !== this) {
            throw new Error('collectStart/collectEnd was called out of order.');
        }
    }
    clearDependencies() {
        this.dependencies.forEach(dep => dep.unsubscribe(this.markDirty));
        this.dependencies.clear();
    }
    collect() {
        const signal = DependencyContext.collectionStack.at(-1);
        if (signal) {
            signal.dependencies.add(this.event.subscribable);
            this.event.subscribe(signal.markDirty);
        }
    }
    dispose() {
        this.clearDependencies();
        this.event.clear();
        this.owner = null;
    }
    async toPromise() {
        let promises = DependencyContext.consumePromises();
        do {
            await Promise.all(promises.map(handle => handle.promise));
            this.invokable();
            promises = DependencyContext.consumePromises();
        } while (promises.length > 0);
        return this.invokable;
    }
}
DependencyContext.collectionSet = new Set();
DependencyContext.collectionStack = [];
DependencyContext.promises = [];
