// @motion-canvas/core/lib/app/ImageExporter.js
import { BoolMetaField, EnumMetaField, NumberMetaField, ObjectMetaField, } from '../meta';
import { clamp } from '../tweening';
import { FileTypes } from './presets';
const EXPORT_FRAME_LIMIT = 256;
const EXPORT_RETRY_DELAY = 1000;
/**
 * Image sequence exporter.
 */
export class ImageExporter {
    constructor(logger) {
        this.logger = logger;
        this.name = 'image sequence';
        this.frameLookup = new Map();
        this.frameCounter = 0;
        this.projectName = 'unknown';
        this.quality = 1;
        this.fileType = 'image/png';
        this.groupByScene = false;
        if (import.meta.hot) {
            import.meta.hot.on('motion-canvas:export-ack', ({ frame }) => {
                this.frameLookup.get(frame)?.();
            });
        }
    }
    meta() {
        const meta = new ObjectMetaField(this.name, {
            fileType: new EnumMetaField('file type', FileTypes),
            quality: new NumberMetaField('quality', 100).setRange(0, 100),
            groupByScene: new BoolMetaField('group by scene', false),
        });
        meta.fileType.onChanged.subscribe(value => {
            meta.quality.disable(value === 'image/png');
        });
        return meta;
    }
    async configure(settings) {
        const options = settings.exporter.options;
        this.projectName = settings.name;
        this.quality = clamp(0, 1, options.quality / 100);
        this.fileType = options.fileType;
        this.groupByScene = options.groupByScene;
    }
    async start() {
        this.frameLookup.clear();
    }
    async handleFrame(canvas, frame, sceneFrame, sceneName, signal) {
        if (this.frameLookup.has(frame)) {
            this.logger.warn(`Frame no. ${frame} is already being exported.`);
            return;
        }
        if (import.meta.hot) {
            while (this.frameCounter > EXPORT_FRAME_LIMIT) {
                await new Promise(resolve => setTimeout(resolve, EXPORT_RETRY_DELAY));
                if (signal.aborted) {
                    return;
                }
            }
            this.frameCounter++;
            this.frameLookup.set(frame, () => {
                this.frameCounter--;
                this.frameLookup.delete(frame);
            });
            import.meta.hot.send('motion-canvas:export', {
                frame,
                sceneFrame,
                data: canvas.toDataURL(this.fileType, this.quality),
                mimeType: this.fileType,
                subDirectories: this.groupByScene
                    ? [this.projectName, sceneName]
                    : [this.projectName],
                groupByScene: this.groupByScene,
            });
        }
    }
    async stop() {
        while (this.frameCounter > 0) {
            await new Promise(resolve => setTimeout(resolve, EXPORT_RETRY_DELAY));
        }
    }
}
