// @motion-canvas/core/lib/meta/ExporterMetaFile.js
import { MetaField } from './MetaField';
import { EnumMetaField } from './EnumMetaField';
import { ValueDispatcher } from '../events';
/**
 * Represents the exporter configuration.
 */
export class ExporterMetaField extends MetaField {
    /**
     * Triggered when the nested fields change.
     *
     * @eventProperty
     */
    get onFieldsChanged() {
        return this.fields.subscribable;
    }
    get options() {
        return this.optionFields[this.current];
    }
    get exporter() {
        return this.exporters[this.current];
    }
    constructor(name, exporters, current = 0) {
        const optionFields = exporters.map(exporter => exporter.meta());
        const exporterField = new EnumMetaField('exporter', exporters.map(exporter => ({
            value: exporter.name,
            text: exporter.name,
        })), exporters[current].name);
        super(name, {
            name: exporterField.get(),
            options: optionFields[current].get(),
        });
        this.exporters = exporters;
        this.current = current;
        this.type = Object;
        this.handleChange = () => {
            const value = this.exporterField.get();
            const index = Math.max(this.exporters.findIndex(exporter => exporter.name === value), 0);
            if (this.current !== index) {
                this.options.onChanged.unsubscribe(this.handleChange);
                this.current = index;
                this.options.onChanged.subscribe(this.handleChange, false);
                this.fields.current = [this.exporterField, this.options];
            }
            this.value.current = {
                name: this.exporterField.get(),
                options: this.options.get(),
            };
        };
        this.exporterField = exporterField;
        this.exporterField.onChanged.subscribe(this.handleChange, false);
        this.exporterField.disable(optionFields.length < 2).space();
        this.optionFields = optionFields;
        this.optionFields[current].onChanged.subscribe(this.handleChange, false);
        this.fields = new ValueDispatcher([this.exporterField, this.options]);
    }
    set(value) {
        this.exporterField.set(value.name);
        this.options.set(value.options);
    }
    serialize() {
        return {
            name: this.exporterField.serialize(),
            options: this.options.serialize(),
        };
    }
    clone() {
        return new this.constructor(this.name, this.exporters, this.current);
    }
}
